### PR TITLE
CASMCMS-7676 Enable loading of recipes with template-dicts

### PR DESCRIPTION
**Pull request by ecozzi-hpe**
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

This PR updates ims-load-artifacts to be able to add recipes to IMS that contain template-dictionaries. 

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_

yes

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CASMCMS-7676](https://jira-pro.its.hpecorp.net:8443/browse/CASMCMS-7676)
* [ ] Merge after ims-python-helper changes (https://github.com/Cray-HPE/ims-python-helper/pull/4)

## Testing

_List the environments in which these changes were tested._

### Tested on:

  * `mug`

### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

Created a test recipe helm chart based on the updated cray-product-install-chart and ims-load-artifacts container. Installed this test recipe helm chart on mug. Saw that the recipe was added to IMS as expected, including that all templating during the helm install occurred as expected. Verified that building the recipe using IMS templated the recipe as expected.

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)? No
- Were continuous integration tests run? If not, why? No
- Was upgrade tested? If not, why? Yes
- Was downgrade tested? If not, why? Yes
- Were new tests (or test issues/Jiras) created for this change? No

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable
- [ ] Do not merge until CSMv2.5 code is able to be merged

### CODE REVIEW

**Comment by ecozzi-hpe** [comment]
Builds will fail until the version of IMS-python-helper is updated.

File: ims_load_artifacts/load_artifacts.py [hunk 1]
```diff
@@ -417,21 +423,50 @@ def __call__(self, manifest_data):
...
+            manifest_data = yaml.safe_load(inf)
+            LOGGER.info("Manifest Data %s", manifest_data)
+
+        try:
+            return {
+                "1.0.0": _ImsLoadArtifacts_v1_0_0()
+            }[manifest_data["version"]](manifest_data)
+        except KeyError:
+            raise ImsLoadArtifactsBaseException(
+                f"Cannot process manifest.yaml due to unsupported or missing manifest version {manifest_data}") \
```
Comment (jsollom-hpe) [inline ims_load_artifacts/load_artifacts.py:465]: Did you want to output the entire manifest_data here or just manifest_data["version"]?

**Review by jsollom-hpe — APPROVED** [review]

File: constraints.txt [hunk 2]
```diff
@@ -4,7 +4,7 @@ certifi==2019.11.28
 chardet==3.0.4
 docutils==0.14
 idna==2.8
-ims-python-helper==2.7.29
+ims-python-helper==0.4.68
```
[the remark this answers is not in the document]
Comment (ecozzi-hpe) [reply constraints.txt:7]: This PR requires a new version of ims-python-helper. But, that version of ims-python-helper is also not yet released/merged. In order to test, I was forced to pin the version of ims-python-helper to this development version. Once ims-python-helper is merged to master and released, the version in this constraints file will need to be updated.

[the remark this answers is not in the document]
Comment (ecozzi-hpe) [reply constraints.txt:7]: - [ ] Pin ims-python-helper to release version

File: ims_load_artifacts/load_artifacts.py (hunk 1 above)
Comment (ecozzi-hpe) [reply ims_load_artifacts/load_artifacts.py:465]: The entire manifest data. The version could either blank or an unsupported version. Reporting the whole manifest gives the most information and allows the user to understand the error.

File: ims_load_artifacts/load_artifacts.py [hunk 3]
```diff
@@ -247,9 +252,10 @@ def load_recipe(self, ih, recipe_name, recipe_data):
             linux_distribution = recipe_data["linux_distribution"]
             link = recipe_data["link"]
             recipe_path = self.download_artifact(link, md5sum)
+            template_dictionary = recipe_data.get("template_dictionary", [])
 
             try:
-                return ih.recipe_upload(recipe_name, recipe_path, linux_distribution)
+                return ih.recipe_upload(recipe_name, recipe_path, linux_distribution, template_dictionary)
```
Comment (rkleinman-hpe) [inline ims_load_artifacts/load_artifacts.py:258]: I believe this also means this change is backwards incompatible, requiring a major version rev.

**Review by rkleinman-hpe — CHANGES_REQUESTED** [review]

File: ims_load_artifacts/load_artifacts.py (hunk 3 above)
Comment (ecozzi-hpe) [reply ims_load_artifacts/load_artifacts.py:258]: I don't know that I agree. The `template_dictionary` value is defaulted it if it is not found in the recipe_data. The version of ims-python-helper is pinned to support this parameter being passed. Not sure how this is a breaking change.

Comment (rkleinman-hpe) [reply ims_load_artifacts/load_artifacts.py:258]: Now that the ims python helper is backwards compatible, I agree.

**Review by mharding-hpe — APPROVED** [review]

**Pull request closed without merging**